### DISCUSSION TITLE
Using the current time observation when simulating to do calculations in the model text string

### DISCUSSION THREAD

**Discussion by jfndiaye** [nlmixrdevelopment/RxODE#489] [Q&A] [answered]
Hi everyone,

Is there a way to use the current time observation (defined with `add.sampling()`) when simulating, to calculate some algebraic variables including time for example in the model text string with RxODE?

Thank you,

**Comment by mattfidler** [comment, marked as answer]
You can use `time` or `t` to calculate time-based variables.  You can also use `tad()` and other such variables.  They are documented in the `rxode2` syntax section:

https://nlmixr2.github.io/rxode2/articles/rxode2-syntax.html#supported-functions

https://nlmixr2.github.io/rxode2/articles/rxode2-syntax.html#reserved-keywords

They should be the same in `RxODE`, though `rxode2` may have diverged slightly.

**Comment by jfndiaye** [comment]
Thank you for your response.

I'm dealing with an issue.
Suppose `var1` and `var2` are two variables defined by two odes, and `av` stores the mean of `var1` at every 24 time steps. The ode that defines `var2` also takes` av` as input.

To implement `av`, I've written the following code which avoid the use of the modulo operator `%%`, the loop `for` and the `mean` function, which seem not to be supported. Now the syntax of `while` seems also not to work as well as `var1[i]`.
```
if (t==0){
  av = av_ini;
}
else{
  if (round(t/24)*24==t){
    som = 0;
    i = t-24;
    while (i<t){
      som = som + var1[i];
      i=i+1;
    }
    av = som/24;
  }
}
```
If you have any suggestion on how to solve this problem. Thank you.

**Comment by mattfidler** [comment]
I am not really sure if you want a mean or not.  In ODE solving you can query points outside of your original time-point mesh, so it may be impossible.  

If you are not concerned about the mesh size and think that the number of samples to create the mean is irrelevant you could do something like this (using Welford's method for mean but skipping the variance)

``` r
library(rxode2)
#> rxode2 2.0.14 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
mod1 <- function() {
  ini({
    # central
    KA=2.94E-01
    CL=1.86E+01
    V2=4.02E+01
    # peripheral
    Q=1.05E+01
    V3=2.97E+02
    # effects
    Kin=1
    Kout=1
    EC50=200
    avIni <- 0
  })
  model({
    if (t == 0) {
      av <- avIni
      runN <- 0
      runMean <- 0
    }
    C2 <- centr/V2
    C3 <- peri/V3
    runN <- runN + 1
    runMean <- runMean + (C2 - runMean) / runN
    if (t != 0 && round(t / 24) == t / 24)  {
      av <- runMean
      # not sure if you wish to reset too
      runN <- 0
      runMean <- 0
    }
    d/dt(depot) <- -KA*depot + av
    d/dt(centr) <- KA*depot - CL*C2 - Q*C2 + Q*C3
    d/dt(peri)  <- Q*C2 - Q*C3
    eff(0) <- 1
    d/dt(eff)   <- Kin - Kout*(1-C2/(EC50+C2))*eff
  })
}

e <- et(amt=100, cmt="depot", ii=12, until=24
        * 4) %>% et(0, 24 * 8, by=0.5)

# This solves with a mean based on 24 hour interval which is adaptive
# depending on how many times the ODE solving engine was called during
# that 24 hour period.  With stiff problems the number of calls may be
# higher in certain regions than in others.

# If that is acceptable and a measure of the mean tendency in the 24
# hour period is what you want for your problem this would likely
# work:

f <- rxSolve(mod1, e)
plot(f, C2, av)
```

![](https://i.imgur.com/9r6c5Si.png)

``` r


# Note that the number of sampling times is different (but similar)
# between each sampling period, as shown in the following plot:
plot(f, av, runN)
```

![](https://i.imgur.com/RRMzmCQ.png)

<sup>Created on 2023-11-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>


You are using approaching your problem as if `rxode2` syntax is the same as R syntax.  While it seems that way at first glance, it is its own mini-language.  The model is called only at a single time and cannot be currently changed to R vectors and call some arbitrary R function like `mean()`.

**Comment by jfndiaye** [comment]
Yes that's what I wanted to do, it works well.

Thank you!
